src/hooks/useNodeViews.tsx

```diff
@@ -0,0 +1,75 @@
+import React, { ReactPortal, useCallback, useMemo, useState } from "react";
+import { createPortal } from "react-dom";
+
+import {
+  PORTAL_REGISTRY_ROOT_KEY,
+  PortalRegistryContext,
+  PortalRegistryKey,
+} from "../contexts/PortalRegistryContext.js";
+import {
+  ReactNodeViewConstructor,
+  RegisterElement,
+  createReactNodeViewConstructor,
+} from "../nodeViews/createReactNodeViewConstructor.js";
+
+type Props = {
+  portals: Record<PortalRegistryKey, ReactPortal[]>;
+};
+
+function NodeViews({ portals }: Props) {
+  const rootPortals = portals[PORTAL_REGISTRY_ROOT_KEY];
+
+  return (
+    <PortalRegistryContext.Provider value={portals}>
+      {rootPortals}
+    </PortalRegistryContext.Provider>
+  );
+}
+
+export function useNodeViews(
+  nodeViews: Record<string, ReactNodeViewConstructor>
+) {
+  const [portals, setPortals] = useState(
+    {} as Record<PortalRegistryKey, ReactPortal[]>
+  );
+
+  const registerPortal: RegisterElement = useCallback(
+    (registrationKey, child, container, key) => {
+      const portal = createPortal(child, container, key);
+      setPortals((oldPortals) => {
+        const oldChildPortals = oldPortals[registrationKey] ?? [];
+        const newChildPortals = oldChildPortals.concat(portal);
+        return {
+          ...oldPortals,
+          [registrationKey]: newChildPortals,
+        };
+      });
+
+      return () => {
+        setPortals((oldPortals) => {
+          const oldChildPortals = oldPortals[registrationKey] ?? [];
+          const newChildPortals = oldChildPortals.filter((p) => p !== portal);
```

Comment:
Should we care about omitting this from the result if it's now zero-length?

Reply:
Hm. I suppose it is a memory leak opportunity. We can do that!